BeardedSpice/Tabs/QuodLibetTabAdapter.m

```diff
@@ -0,0 +1,382 @@
+            return _track;
+        }
+        
+        // read track status from current file
+        NSError *error = nil;
+        NSString *current = [[NSString alloc] initWithContentsOfFile:[QL_CURRENT stringByExpandingTildeInPath]
+                                              encoding:NSUTF8StringEncoding error:&error];
+        if (error) {
+            NSLog(@"Failed to read QuodLibet current track info: %@", [error localizedDescription]);
+            return nil;
+        }
+        
+        _track = [BSTrack new];
+        NSString *filename;
+
+        NSArray *lines = [current componentsSeparatedByCharactersInSet:[NSCharacterSet newlineCharacterSet]];
+        for (NSString *line in lines) {
+            NSRange delim = [line rangeOfString:@"="];
+            if (delim.location != NSNotFound) {
+                NSString *key = [line substringToIndex:delim.location];
+                NSString *value = [line substringFromIndex:delim.location + 1];
+                
+                if ([key isEqualToString:QL_TRACK_TITLE]) {
+                    _track.track = value;
+                } else if ([key isEqualToString:QL_TRACK_ARTIST]) {
+                    _track.artist = value;
+                } else if ([key isEqualToString:QL_TRACK_ALBUM]) {
+                    _track.album = value;
+                } else if ([key isEqualToString:QL_TRACK_FILENAME]) {
+                    filename = value;
+                } else if ([key isEqualToString:QL_TRACK_LENGTH]) {
+                    NSTimeInterval length = [value doubleValue];
+                    NSTimeInterval position = length * _progress;
+                    _track.progress = [NSString stringWithFormat:@"%@ of %@",
+                                             FormatNSTimeInterval(position),
+                                             FormatNSTimeInterval(length)];
+                }
+            }
+        }
+        
+        if ([NSString isNullOrEmpty:_track.track]) {
+            // No title field for this track, fall back to basename.
+            // QuodLibet's "current" file is guaranteed to have a ~filename tag, we can
+            // rely on that here. The output format here, using the base
+            // filename with localised `[Unknown]` postfix, echoes how QL generates
+            // a track title for such tracks.
+            NSString *basename = [filename lastPathComponent];
+            if ([NSString isNullOrEmpty:basename]) {
+                basename = filename;
+            }
+            NSString *unknown = NSLocalizedString(@"Unknown", @"QuodLibetTabAdapter");
+            _track.track = [NSString stringWithFormat:@"%@ [%@]", basename, unknown];
+        }
+
+        return _track;
+    }
+    return nil;
+}
+
+- (BOOL)isPlaying{
+    [self loadQuodLibetState];
+    return _status == QuodLibetStatusPlaying;
+}
+
+#pragma mark QuodLibet remote control methods
+
+- (void)loadQuodLibetState{
+    if (_status != QuodLibetStatusUnknown) {
+        return;
```

Comment:
@mjpieters Is it legal at all? :)

Reply:
there is a feeling that the method will work completely only once...

Reply:
BeardedSpice creates new instances of each native app tab adapter for every interaction. `_status` is caching the QL status only for the lifetime of each instance.

So, all this works fine for me, I've been using this code for a few days now without issue. :-)

Reply:
"BeardedSpice creates new instances of each native app tab adapter for every interaction. " Hm. Sorry, I forgot that is different here :)

Reply:
I initially had more cache invalidation logic with a timestamp but found that the cache was gone every time, until I realised that the lifetime of any adapter instance is really short.

Per interaction we do want to cache this data as the QL named pipe API is relatively slow.